src/core/namespace.c

```diff
@@ -153,162 +149,117 @@ static const TargetMount protect_system_strict_table[] = {
         { "/root",      READWRITE,      true  },      /* ProtectHome= */
 };
 
-static void set_bind_mount(BindMount *p, char *path, MountMode mode, bool ignore) {
-        p->path = path;
-        p->mode = mode;
-        p->ignore = ignore;
-}
-
-static int append_one_mount(BindMount **p, const char *root_directory,
-                            const char *path, MountMode mode, bool ignore) {
-        char *lpath;
+static const char *bind_mount_path(const BindMount *p) {
         assert(p);
 
-        lpath = prefix_root(root_directory, path);
-        if (!lpath)
-                return -ENOMEM;
+        /* Returns the path of this bind mount. If the malloc()-allocated ->path_buffer field is set we return that,
+         * otherwise the stack/static ->path field is returned. */
 
-        set_bind_mount((*p)++, lpath, mode, ignore);
-        return 0;
+        return p->path_malloc ?: p->path_const;
 }
 
-static int append_mounts(BindMount **p, char **strv, MountMode mode) {
+static int append_access_mounts(BindMount **p, char **strv, MountMode mode) {
         char **i;
 
         assert(p);
 
-        STRV_FOREACH(i, strv) {
-                bool ignore = false;
-                char *path;
+        /* Adds a list of user-supplied READWRITE/READONLY/INACCESSIBLE entries */
 
-                if (IN_SET(mode, INACCESSIBLE, READONLY, READWRITE) && startswith(*i, "-")) {
-                        (*i)++;
-                        ignore = true;
+        STRV_FOREACH(i, strv) {
+                bool ignore = false, has_prefix = true;
+                const char *e = *i;
+
+                /* Look for any prefixes */
+                for (;;) {
+                        if (startswith(e, "-")) {
```

Comment:
Hm, I think we shouldn't allow "+-blah". I think "+" should be always directly before the path. Also, duplicates should not be allowed.
``` C
      bool ignore = false, needs_prefix = false;
      if (startswith(e, "-")) {
           e++;
           ignore = true;
      }
      if (startswith(e, "+")) {
          e++;
          needs_prefix = true;
      }
```